sdks/go/examples/native_wordcap/nativepubsubio/native.go

```diff
@@ -0,0 +1,247 @@
+			return err
+		}
+		r.client = client
+	}
+	return nil
+}
+
+var messageTimeout time.Duration = 5 * time.Second
+
+// ProcessElement initializes a PubSub client if one has not been created already, reads from the PubSub subscription,
+// and emits elements as it reads them. If no messages are available, the DoFn will schedule itself to resume processing
+// later. If polling the subscription returns an error, the error will be logged and the DoFn will not reschedule itself.
+func (r *pubSubRead) ProcessElement(ctx context.Context, bf beam.BundleFinalization, rt *sdf.LockRTracker, _ []byte, emit func(beam.EventTime, []byte)) (sdf.ProcessContinuation, error) {
+	// Register finalization callback
+	bf.RegisterCallback(5*time.Minute, func() error {
+		for _, m := range r.processedMessages {
+			m.Ack()
+		}
+		r.processedMessages = nil
+		return nil
+	})
+
+	for {
+		ok := rt.TryClaim(r.Subscription)
+		if !ok {
+			return sdf.ResumeProcessingIn(5 * time.Second), nil
+		}
+		sub := r.client.Subscription(r.Subscription)
+		ctx, cFn := context.WithCancel(ctx)
+
+		// Because emitters are not thread safe and synchronous Receive() behavior
+		// is deprecated, we have to collect messages in a goroutine and pipe them
+		// out through a channel.
+		messChan := make(chan *pubsub.Message, 1)
+		go func(sendch chan<- *pubsub.Message) {
+			err := sub.Receive(ctx, func(ctx context.Context, m *pubsub.Message) {
+				messChan <- m
+			})
+			if (err != nil) && (err != context.Canceled) {
+				log.Errorf(ctx, "error reading from PubSub: %v, stopping processing", err)
+				cFn()
+				close(messChan)
+			}
+		}(messChan)
+
+		timeout := time.NewTimer(messageTimeout)
+		for {
+			select {
+			case m, ok := <-messChan:
+				if !ok {
+					log.Debug(context.Background(), "stopping bundle processing")
+					return sdf.StopProcessing(), nil
+				}
+				r.processedMessages = append(r.processedMessages, m)
+				emit(beam.EventTime(m.PublishTime.UnixMilli()), m.Data)
+				if !timeout.Stop() {
+					<-timeout.C
+				}
+				timeout.Reset(messageTimeout)
+			case <-timeout.C:
+				log.Debugf(context.Background(), "cancelling receive context, scheduling resumption")
+				cFn()
+				return sdf.ResumeProcessingIn(10 * time.Second), nil
+			}
+		}
+	}
+}
+
+// NativeRead reads messages from a PubSub topic in a streaming context, outputting
+// received messages as a PCollection of byte slices. If the provided subscription
+// name exists for the given topic, the DoFn will read from that subscription; otherwise,
+// a new subscription with the given subscription name will be created and read from.
+//
+// This feature is experimental and subject to change, including its behavior and function signature.
+// Please use the cross-language implementation Read() instead.
+func NativeRead(s beam.Scope, project, topic, subscription string) beam.PCollection {
+	s = s.Scope("pubsubio.NativeRead")
+
+	psRead, err := newPubSubRead(context.Background(), project, topic, subscription)
```

Comment:
Idiomatic go would be to accept the context as the first parameter and then use that, which signals to the user that some RPCs may happen, and allow them to do whatever machinations they desire to the context.

Idiomatic Go has context.Background() only called by end binary authors, rather than at arbitrary library levels & using context.TODO() to signal where additional plumbing is required when a caller doesn't have a context parameter itself.

Reply:
Gotcha. Changed